### ISSUE TITLE
Parser allows JSON schemas to participate in type specialization, expression and inheritance

### ISSUE THREAD

**Issue by deiteris** [aml-org/amf#1599]
Your issue may already be reported! Please search on Github issues before creating one.

* **I'm submitting a ...**
  - [x] bug report
  - [ ] feature request
  - [ ] question
  - [ ] other

* **What is the current behavior?**

When working with RAML types and JSON schemas, the parser allows JSON schema types to participate in type inheritance and specialization and in type expressions while the spec is strictly against this.

>A RAML processor MUST NOT allow types that define an XML or JSON schema to participate in type inheritance or specialization, or effectively in any [type expression](https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md#type-expressions). Therefore, you SHALL NOT define sub-types of these types to declare new properties, add restrictions, set facets, or declare facets.

https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md#using-xml-and-json-schemas

* **If the current behavior is a bug, please provide the minimal steps to reproduce with AMF code. Please don't upload AMF code mixed with yours.**

The following RAML reproduces the issue.

```yaml
#%RAML 1.0

title: Pets API
version: 1.0

types:
    MyJsonType: |
        {
            "$schema": "http://json-schema.org/draft-04/schema#",
            "type": "string",
            "minLength": 0
        }
    MyRamlType:
        type: integer
        minimum: 100
    JsonRamlUnionType: MyJsonType | MyRamlType # Should not be valid
    JsonArrayType: MyJsonType[] # Should not be valid
    JsonInheritedType:
        type: MyJsonType
        maxLength: 100 # Should not be valid
```

* **What is the expected behavior?**

Even though the parser produces correct results (all constraints, type inheritance and specialization are preserved), these manipulations should be forbidden according to the RAML spec, and the parser should produce validation errors.

* **Please tell us about your environment:**
  
  - AMF Version: 5.0.12
  - AMF Distribution: JS 
  - Operating System: Windows 10 Pro 21H2

**Comment by hghianni** [comment]
Hi @deiteris, there is a bug when the JSON schema is not external. That's the reason why you are not seeing any validation errors. We throw a warning in this cases. You may wrap the report and map the warning to a violation if you prefer.  I created an internal [issue](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE0000176OHzYAM/view) to fix it. Thank you!

**Comment by deiteris** [comment]
Hi @hghianni, you mean if I include JSON schema from a file, I should get validation warnings? I extracted the JSON schema into a file, but still don't get any validation messages.

```yaml
#%RAML 1.0

title: Pets API
version: 1.0

types:
    MyJsonType: !include MyJsonType.json
    MyRamlType:
        type: integer
        minimum: 100
    JsonRamlUnionType: MyJsonType | MyRamlType
    JsonArrayType: MyJsonType[]
    JsonInhertedType:
        type: MyJsonType
        maxLength: 100
```

The code I use to parse and validate:

```js
import amf from 'amf-client-js'

async function parseRaml(path) {
    const renderOptions = new amf.RenderOptions().withoutSourceMaps().withCompactUris().withoutPrettyPrint();
    const client = amf.RAMLConfiguration.RAML10().withRenderOptions(renderOptions).baseUnitClient();
    const result = await client.parse(`file://${path}`)
    const transformed = client.transform(result.baseUnit, amf.PipelineId.Default)
    const api = client.renderGraphToBuilder(transformed.baseUnit, new amf.org.yaml.builder.JsOutputBuilder())

    const validationResult = await client.validate(result.baseUnit)
    const report = result.merge(validationResult)
    console.log(report.toString())
}

(async () => {
    await parseRaml('test.raml')
})();
```

The output is as follows:

```
Model: file://test.raml
Conforms: true
Number of results: 0
```

**Comment by hghianni** [comment]
If you include the JSON in a type and then you inherit from that type, it uses the same malfunctioning mechanism to detect inheritance as in your first example. Validations will appear if you include the JSON directly where you want to inherit from it:

```
#%RAML 1.0

title: Pets API
version: 1.0

types:
  PersonArray:
    type: array
    items:
      !include MyJsonType.json
```

**Comment by deiteris** [comment]
Okay, but this seems to be working only if JSON schema is a subtype of `object` and `array` types and only without type comprehension. Moreover, the same RAML will not trigger validation warning if JSON schema is declared as RAML type:
```yaml
#%RAML 1.0

title: Pets API
version: 1.0

types:
    MyJsonType: !include MyJsonType.json
    PersonArray:
        type: array
        items: MyJsonType # Will not trigger http://a.ml/vocabularies/amf/parser#json-schema-inheritance
    PersonArrayInclude:
        type: array
        items: !include MyJsonType.json # Triggers http://a.ml/vocabularies/amf/parser#json-schema-inheritance
    PersonObject:
        type: object
        properties: 
            prop: MyJsonType # Will not trigger http://a.ml/vocabularies/amf/parser#json-schema-inheritance
```

**Comment by arielmirra** [comment]
Hi @deiteris! This has now been solved and the fix will be available in amf version 5.2.6 (on april release). 

The only catch is we couldn't solve the case for the `JsonInhertedType` now because that requires a refactor in inheritance detection (that is currently ongoing), but all other cases where a JSON Schema is used in type expression now throw a warning. 

Thank you!
